content/blog/2025-05-13-bril-concurrency-extension.md

```diff
@@ -0,0 +1,163 @@
+
+To simulate true concurrency, we leveraged Deno’s Web Workers. Each spawn:
+
+1.  Allocates a unique thread ID and a corresponding promise resolver.
+    
+2.  Spins up a new Worker running brili\_worker.ts, passing it the full Bril program, function name, arguments, **and a proxy to the shared Heap**.
+    
+3.  The worker executes runBrilFunction(...) in its own isolate and signals completion via postMessage.
+    
+4.  The main isolate bridges heap operations through an RPC-style HeapProxy, intercepting alloc/load/store/freefrom workers and dispatching them on the master heap.
+
+
+This architecture faithfully exposes nondeterministic interleaving and shared-memory errors (e.g., data races), but at the cost of heavy message-passing overhead.
+
+3\. Thread-Escape Analysis Pass
+-------------------------------
+
+We wrote examples/escape.py, a Python script that performs a conservative, intraprocedural escape analysis:
+
+1.  **Seed escapes**: Pointers passed to spawn, returned via ret, stored into other pointers, or passed to external calls are marked as escaping.
+    
+2.  **Propagation**: Through ptradd and load chains, any derived pointer from an escaping pointer also escapes.
+    
+3.  **Reporting**: The pass reports per-function and global statistics: the fraction of allocations that remain thread-local.
+    
+
+This simple pass identifies heap objects that never cross thread boundaries, opening opportunities to Elide synchronization or use stack allocation in a future optimizer.
+
+4\. Benchmark Suites
+--------------------
+
+We developed two complementary benchmark suites:
+
+*   **Correctness Suite** (benchmarks/concurrency/\*.bril): A set of small tests covering spawn/join arity errors, double joins, shared counters, parallel sums, and array writers. We validated behavior under both the stubbed and worker-based interpreters using turnt.
+    
+*   **Performance Suite** (benchmarks/concurrency/perf/\*.bril): Larger, data-parallel workloads measuring speed-up with workers:
+    
+    *   perf\_par\_sum\_100k.bril: Summing 1..100 000 split across two threads. 100 k elements produce ~20× speed-up over sequential.
+        
+    *   perf\_matmul\_split.bril: Matrix multiplication on 100×100 matrices with row-range split; yields ~10× speed-up.
+        
+    *   perf\_big\_par\_sum\_10M.bril: Summing 1..10 000 000, which instead suffers a slowdown due to RPC overhead.
+        
+
+5\. Empirical Evaluation
+------------------------
+
+### 5.1 Correctness
+
+All the correctness benchmarks pass under the worker-enabled interpreter (brili-conc), matching our specification. Error cases in the error directory (arity mismatch, double join, wrong handle type) correctly exit with error codes.
+
+### 5.2 Performance
+
+Below are the wall-clock times measured for three representative benchmarks, comparing the sequential interpreter (using --no-workers) against our concurrent version (with true Deno Web Workers and an RPC-based shared heap).
```

Comment:
Can you say something about your experimental setup (hardware/OS/Deno runtime) and how you took measurements (using Unix `time`, or Hyperfine with how many replica)?